File: src/main/java/com/github/jaksonlin/pitestintellij/annotations/AnnotationValidator.java
package com.github.jaksonlin.pitestintellij.annotations;

import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

public class AnnotationValidator {
    private final AnnotationSchema schema;

    public AnnotationValidator(AnnotationSchema schema) {
        this.schema = schema;
    }

    public static abstract class ValidationResult {
        public static class Valid extends ValidationResult {
        }

        public static class Invalid extends ValidationResult {
            private final List<String> errors;

            public Invalid(List<String> errors) {
                this.errors = errors;
            }

            public List<String> getErrors() {
                return errors;
            }
        }
    }

    public ValidationResult validate(Map<String, Object> annotationValues) {
        List<String> errors = new ArrayList<>();

        // Check required fields
        schema.getFields().stream()
                .filter(AnnotationFieldConfig::isRequired)
                .forEach(field -> {
                    if (!annotationValues.containsKey(field.getName()) || annotationValues.get(field.getName()) == null) {
                        errors.add("Missing required field: " + field.getName());
                    }
                });

        // Validate field types
        annotationValues.forEach((name, value) -> schema.getFields().stream()
                .filter(field -> field.getName().equals(name))
                .findFirst()
                .ifPresent(field -> {
                    String error = validateField(field, value);
                    if (error != null) {
                        errors.add(error);
                    }
                }));

        return errors.isEmpty() ? new ValidationResult.Valid() : new ValidationResult.Invalid(errors);
    }

    @Nullable
    private String validateField(AnnotationFieldConfig field, @Nullable Object value) {
        if (value == null) return null;

        // First validate the type
        String typeError = validateType(field, value);
        if (typeError != null) return typeError;

        // Then validate non-empty if required
        String emptyError = validateNonEmpty(field, value);
        if (emptyError != null) return emptyError;

        // Then validate against the validation rules if they exist
        FieldValidation validation = field.getValidation();
        if (validation != null) {
            if (value instanceof String) {
                String error = validateStringValue(field.getName(), (String) value, validation);
                if (error != null) return error;
            } else if (value instanceof List<?>) {
                // Validate list content
                String contentError = validateListContent(field.getName(), (List<?>) value, validation);
                if (contentError != null) return contentError;

                // Validate list length
                String lengthError = validateListLength(field.getName(), (List<?>) value, validation);
                if (lengthError != null) return lengthError;
            } else {
                // Should not happen
                return "Unsupported field type for " + field.getName() + ": " + value.getClass().getSimpleName();
            }
        }

        return null;
    }

    @Nullable
    private String validateNonEmpty(AnnotationFieldConfig field, Object value) {
        FieldValidation validation = field.getValidation();
        if (validation != null && !validation.isAllowEmpty()) {
            if (value instanceof String) {
                if (((String) value).trim().isEmpty()) {
                    return "Field " + field.getName() + " cannot be empty";
                }
            } else if (value instanceof List<?>) {
                if (((List<?>) value).isEmpty()) {
                    return "Field " + field.getName() + " cannot be empty";
                }
            }
        }
        return null;
    }

    @Nullable
    private String validateType(AnnotationFieldConfig field, Object value) {
        switch (field.getType()) {
            case STRING:
                if (!(value instanceof String)) return "Field " + field.getName() + " must be a String";
                else return null;
            case STRING_LIST:
                if (value instanceof List<?>) {
                    for (Object item : (List<?>) value) {
                        if (!(item instanceof String)) {
                            return "Field " + field.getName() + " must be a list of Strings";
                        }
                    }
                    return null;
                } else return "Field " + field.getName() + " must be a List";
            default:
                return null;
        }
    }

    @Nullable
    private String validateStringValue(
            String fieldName,
            String value,
            FieldValidation validation
    ) {
        if (!validation.isAllowEmpty() && value.trim().isEmpty()) {
            return "Field " + fieldName + " cannot be empty";
        }
        if (!validation.getValidValues().isEmpty() && !validation.isAllowCustomValues()) {
            boolean isValid = false;
            switch (validation.getMode()) {
                case EXACT:
                    isValid = validation.getValidValues().contains(value);
                    break;
                case CONTAINS:
                    for (String candidate : validation.getValidValues()) {
                        if (value.toLowerCase().contains(candidate.toLowerCase())) {
                            isValid = true;
                            break;
                        }
                    }
                    break;
            }
            if (!isValid) {
                return "Invalid value for " + fieldName + ": " + value + ". Valid values are: " + String.join(", ", validation.getValidValues());
            }
        }
        return null;
    }

    @Nullable
    private String validateListContent(
            String fieldName,
            List<?> value,
            FieldValidation validation
    ) {
        if (!validation.getValidValues().isEmpty() && !validation.isAllowCustomValues()) {
            List<String> invalidValues = value.stream()
                    .filter(String.class::isInstance)
                    .map(String.class::cast)
                    .filter(item -> {
                        switch (validation.getMode()) {
                            case EXACT:
                                return !validation.getValidValues().contains(item);
                            case CONTAINS:
                                return validation.getValidValues().stream()
                                        .noneMatch(candidate -> item.toLowerCase().contains(candidate.toLowerCase()));
                            default:
                                return false;
                        }
                    })
                    .collect(Collectors.toList());

            if (!invalidValues.isEmpty()) {
                return "Invalid values for " + fieldName + ": " + String.join(", ", invalidValues) + ". Valid values are: " + String.join(", ", validation.getValidValues());
            }
        }
        return null;
    }

    @Nullable
    private String validateListLength(
            String fieldName,
            List<?> value,
            FieldValidation validation
    ) {
        if (!validation.isAllowEmpty() && value.isEmpty()) {
            return "Field " + fieldName + " cannot be empty";
        }
        Integer min = validation.getMinLength();
        if (min != null && value.size() < min) {
            return fieldName + " must contain at least " + min + " element" + (min > 1 ? "s" : "");
        }
        Integer max = validation.getMaxLength();
        if (max != null && value.size() > max) {
            return fieldName + " cannot contain more than " + max + " element" + (max > 1 ? "s" : "");
        }
        return null;
    }
}

File: src/main/kotlin/com/github/jaksonlin/pitestintellij/annotations/AnnotationValidator.kt
package com.github.jaksonlin.pitestintellij.annotations

class AnnotationValidator(private val schema: AnnotationSchema) {
    sealed class ValidationResult {
        object Valid : ValidationResult()
        data class Invalid(val errors: List<String>) : ValidationResult()
    }

    fun validate(annotationValues: Map<String, Any>): ValidationResult {
        val errors = mutableListOf<String>()

        // Check required fields
        schema.fields
            .filter { it.required }
            .forEach { field ->
                if (!annotationValues.containsKey(field.name) || annotationValues[field.name] == null) {
                    errors.add("Missing required field: ${field.name}")
                }
            }

        // Validate field types
        annotationValues.forEach { (name, value) ->
            schema.fields.find { it.name == name }?.let { field ->
                validateField(field, value)?.let { error ->
                    errors.add(error)
                }
            }
        }

        return if (errors.isEmpty()) ValidationResult.Valid else ValidationResult.Invalid(errors)
    }

    private fun validateField(field: AnnotationFieldConfig, value: Any?): String? {
        if (value == null) return null

        // First validate the type
        val typeError = validateType(field, value)
        if (typeError != null) return typeError

        // Then validate non-empty if required
        val emptyError = validateNonEmpty(field, value)
        if (emptyError != null) return emptyError

        // Then validate against the validation rules if they exist
        field.validation?.let { validation ->
            when (value) {
                is String -> {
                    validateStringValue(field.name, value, validation)?.let { return it }
                }
                is List<*> -> {
                    // Validate list content
                    validateListContent(field.name, value, validation)?.let { return it }
                    
                    // Validate list length
                    validateListLength(field.name, value, validation)?.let { return it }
                }

                else -> {
                    // Should not happen
                    return "Unsupported field type for ${field.name}: ${value::class.simpleName}"
                }
            }
        }

        return null
    }

    private fun validateNonEmpty(field: AnnotationFieldConfig, value: Any): String? {
        if (!field.validation?.allowEmpty.let { it != null && !it }) {
            when (value) {
                is String -> {
                    if (value.isBlank()) {
                        return "Field ${field.name} cannot be empty"
                    }
                }
                is List<*> -> {
                    if (value.isEmpty()) {
                        return "Field ${field.name} cannot be empty"
                    }
                }
            }
        }
        return null
    }

    private fun validateType(field: AnnotationFieldConfig, value: Any): String? {
        return when (field.type) {
            AnnotationFieldType.STRING -> {
                if (value !is String) "Field ${field.name} must be a String"
                else null
            }
            AnnotationFieldType.STRING_LIST -> {
                when (value) {
                    is List<*> -> {
                        if (value.any { it !is String }) {
                            "Field ${field.name} must be a list of Strings"
                        } else null
                    }
                    else -> "Field ${field.name} must be a List"
                }
            }
            AnnotationFieldType.STATUS -> {
                if (value !is String) "Field ${field.name} must be a String"
                else null
            }
        }
    }

    private fun validateStringValue(
        fieldName: String,
        value: String,
        validation: FieldValidation
    ): String? {
        if (!validation.allowEmpty && value.isBlank()) {
            return "Field $fieldName cannot be empty"
        }
        if (validation.validValues.isNotEmpty() && !validation.allowCustomValues) {
            val isValid = when (validation.mode) {
                ValidationMode.EXACT -> validation.validValues.contains(value)
                ValidationMode.CONTAINS -> validation.validValues.any { candidate -> 
                    value.contains(candidate, ignoreCase = true) 
                }
            }
            if (!isValid) {
                return "Invalid value for $fieldName: $value. Valid values are: ${validation.validValues.joinToString()}"
            }
        }
        return null
    }

    private fun validateListContent(
        fieldName: String,
        value: List<*>,
        validation: FieldValidation
    ): String? {
        if (validation.validValues.isNotEmpty() && !validation.allowCustomValues) {
            val invalidValues = value.filterIsInstance<String>()
                .filter { item ->
                    when (validation.mode) {
                        ValidationMode.EXACT -> !validation.validValues.contains(item)
                        ValidationMode.CONTAINS -> !validation.validValues.any { candidate -> 
                            item.contains(candidate, ignoreCase = true) 
                        }
                    }
                }
            if (invalidValues.isNotEmpty()) {
                return "Invalid values for $fieldName: ${invalidValues.joinToString()}. Valid values are: ${validation.validValues.joinToString()}"
            }
        }
        return null
    }

    private fun validateListLength(
        fieldName: String,
        value: List<*>,
        validation: FieldValidation
    ): String? {
        validation.allowEmpty.let { allowEmpty ->
            if (!allowEmpty && value.isEmpty()) {
                return "Field $fieldName cannot be empty"
            }
        }
        validation.minLength?.let { min ->
            if (value.size < min) {
                return "$fieldName must contain at least $min element${if (min > 1) "s" else ""}"
            }
        }
        validation.maxLength?.let { max ->
            if (value.size > max) {
                return "$fieldName cannot contain more than $max element${if (max > 1) "s" else ""}"
            }
        }
        return null
    }
}